client/Wayland/wlf_disp.c

```diff
@@ -0,0 +1,377 @@
+		return;
+
+	wlf_disp_sendResize(wlfDisp);
+}
+
+wlfDispContext* wlf_disp_new(wlfContext* wlc)
+{
+	wlfDispContext* ret;
+
+	if (!wlc || !wlc->context.settings || !wlc->context.pubSub)
+		return NULL;
+
+	ret = calloc(1, sizeof(wlfDispContext));
+
+	if (!ret)
+		return NULL;
+
+	ret->wlc = wlc;
+	ret->lastSentWidth = ret->targetWidth = wlc->context.settings->DesktopWidth;
+	ret->lastSentHeight = ret->targetHeight = wlc->context.settings->DesktopHeight;
+	PubSub_SubscribeActivated(wlc->context.pubSub, wlf_disp_OnActivated);
+	PubSub_SubscribeGraphicsReset(wlc->context.pubSub, wlf_disp_OnGraphicsReset);
+	PubSub_SubscribeTimer(wlc->context.pubSub, wlf_disp_OnTimer);
+	return ret;
+}
+
+void wlf_disp_free(wlfDispContext* disp)
+{
+	if (!disp)
+		return;
+
+	if (disp->wlc)
+	{
+		PubSub_UnsubscribeActivated(disp->wlc->context.pubSub, wlf_disp_OnActivated);
+		PubSub_UnsubscribeGraphicsReset(disp->wlc->context.pubSub, wlf_disp_OnGraphicsReset);
+		PubSub_UnsubscribeTimer(disp->wlc->context.pubSub, wlf_disp_OnTimer);
+	}
+
+	free(disp);
+}
+
+UINT wlf_disp_sendLayout(DispClientContext* disp, rdpMonitor* monitors, size_t nmonitors)
+{
+	UINT ret = CHANNEL_RC_OK;
+	DISPLAY_CONTROL_MONITOR_LAYOUT* layouts;
+	size_t i;
+	wlfDispContext* wlfDisp = (wlfDispContext*)disp->custom;
+	rdpSettings* settings = wlfDisp->wlc->context.settings;
+	layouts = calloc(nmonitors, sizeof(DISPLAY_CONTROL_MONITOR_LAYOUT));
+
+	if (!layouts)
+		return CHANNEL_RC_NO_MEMORY;
+
+	for (i = 0; i < nmonitors; i++)
+	{
+		layouts[i].Flags = (monitors[i].is_primary ? DISPLAY_CONTROL_MONITOR_PRIMARY : 0);
+		layouts[i].Left = monitors[i].x;
+		layouts[i].Top = monitors[i].y;
+		layouts[i].Width = monitors[i].width;
+		layouts[i].Height = monitors[i].height;
+		layouts[i].Orientation = ORIENTATION_LANDSCAPE;
+		layouts[i].PhysicalWidth = monitors[i].attributes.physicalWidth;
+		layouts[i].PhysicalHeight = monitors[i].attributes.physicalHeight;
+
+		switch (monitors[i].attributes.orientation)
+		{
+			case 90:
+				layouts[i].Orientation = ORIENTATION_PORTRAIT;
+				break;
+
+			case 180:
+				layouts[i].Orientation = ORIENTATION_LANDSCAPE_FLIPPED;
+				break;
+
+			case 270:
+				layouts[i].Orientation = ORIENTATION_PORTRAIT_FLIPPED;
+				break;
+
+			case 0:
+			default:
+				/* MS-RDPEDISP - 2.2.2.2.1:
+				 * Orientation (4 bytes): A 32-bit unsigned integer that specifies the
+				 * orientation of the monitor in degrees. Valid values are 0, 90, 180
+				 * or 270
+				 *
+				 * So we default to ORIENTATION_LANDSCAPE
+				 */
+				layouts[i].Orientation = ORIENTATION_LANDSCAPE;
+				break;
+		}
+
+		layouts[i].DesktopScaleFactor = settings->DesktopScaleFactor;
+		layouts[i].DeviceScaleFactor = settings->DeviceScaleFactor;
+	}
+
+	ret = IFCALLRESULT(CHANNEL_RC_OK, disp->SendMonitorLayout, disp, nmonitors, layouts);
+	free(layouts);
+	return ret;
+}
+
+BOOL wlf_disp_handle_configure(wlfDispContext* disp, int32_t width, int32_t height)
+{
+	if (!disp)
+		return FALSE;
```

Comment:
duplicated check

Reply:
exported in `wlf_disp.h`, no internal static

Reply:
@akallabeth well I don't like the check (as it's not really exported as just used internally, it's present in the header file just for convenience). My point was more that you have exactly the same check 2 lines below.

Reply:
@hardening ok, that duplication was not intentional